etcd/src/main/scala/io/buoyant/etcd/Key.scala

```diff
@@ -0,0 +1,282 @@
+package io.buoyant.etcd
+
+import com.twitter.finagle.{Path, Service}
+import com.twitter.finagle.http._
+import com.twitter.io.Buf
+import com.twitter.util._
+
+case class BackoffsExhausted(key: Path, throwable: Throwable)
+  extends Exception(key.show, throwable)
+
+object Key {
+
+  /*
+   * Helpers for building request params
+   */
+
+  private val Params = Seq.empty[(String, String)]
+
+  private def trueParam(name: String, cond: Boolean): Option[(String, String)] =
+    if (cond) Some(name -> "true") else None
+
+  private def falseParam(name: String, cond: Boolean): Option[(String, String)] =
+    if (!cond) Some(name -> "false") else None
+
+  private[this] val dirParam = ("dir" -> "true")
+  private def dirOrValueParam(v: Option[Buf]): (String, String) = v match {
+    case None => dirParam
+    case Some(buf) =>
+      val Buf.Utf8(v) = buf
+      ("value" -> v)
+  }
+
+  private def ttlParam(ttl: Option[Duration]): Option[(String, String)] = ttl.map {
+    case Duration.Zero => "ttl" -> ""
+    case ttl => "ttl" -> ttl.inSeconds.toString
+  }
+
+}
+
+class Key(key: Path, client: Service[Request, Response]) {
+
+  import Etcd._
+  import Key._
+
+  def path: Path = key
+
+  def key(path: Path): Key = new Key(key ++ path, client)
+  def key(name: String): Key = key(Path.read(name))
+
+  private[this] lazy val uriPath = keysPrefixPath ++ key
+
+  /**
+   * Set the contents of a key.
+   *
+   * If `recursive` is true and this key is a directory, the returned
+   * Node contains the entire tree of children.
+   *
+   * If `wait` is true and `waitIndex` is specified, a response is not
+   * received until the node (or one of its children, if `recursive`
+   * is specified) is updated. If `wait` is true and `waitIndex` is
+   * not specified, a response is not received until the next update
+   * to this node (or its children, if `recursive`).
+   *
+   * If `quorum` is specified, etcd will ensure that the etcd instance
+   * is at quorum with the cluster.
+   */
+  def get(
+    recursive: Boolean = false,
+    wait: Boolean = false,
+    waitIndex: Option[Long] = None,
+    quorum: Boolean = false
+  ): Future[NodeOp] = {
+    val params = Params ++
+      trueParam("recursive", recursive) ++
+      trueParam("quorum", quorum) ++
+      trueParam("wait", wait) ++
+      waitIndex.map("waitIndex" -> _.toString)
+    val req = mkReq(uriPath, params = params)
+    req.headerMap("accept") = MediaType.Json
+    client(req).flatMap { rsp => Future.const(NodeOp.mk(req, rsp, key, params)) }
```

Comment:
tioli: this line is repeated a bunch throughout this file and could be extracted to a private helper method. not a huge deal since it's a one-liner.